Mathlib/RingTheory/Spectrum/Prime/ChevalleyComplexity.lean

```diff
@@ -0,0 +1,893 @@
+Chevalley's theorem with complexity bound. -/
+private def degree : DegreeType n :=
+  toLex (Polynomial.degree ∘ e, ¬ ∃ i, (e i).Monic ∧
+    ∀ j, e j ≠ 0 → (e i).degree ≤ (e j).degree)
+
+@[simp] private lemma ofLex_degree_fst (i) : (ofLex e.degree).fst i = (e i).degree := rfl
+
+private lemma ofLex_degree_snd :
+    (ofLex e.degree).snd ↔
+     ¬ ∃ i, (e i).Monic ∧ ∀ j, e j ≠ 0 → (e i).degree ≤ (e j).degree := .rfl
+
+variable (e) in
+/-- The bound on the degree of the polynomials used to describe the constructible set appearing in
+the case of `C : R → R[X]` of Chevalley's theorem with complexity bound. -/
+private def degBound : ℕ := ∑ i, (e i).degree.succ
+
+variable (e) in
+/-- The bound on the power of the subgroup generated by the coefficients of the polynomials used to
+describe the constructible set appearing in the case of `C : R → R[X]` of Chevalley's theorem with
+complexity bound. -/
+private def powBound : ℕ := e.degBound ^ e.degBound
+
+private lemma powBound_ne_zero : e.powBound ≠ 0 := Nat.pow_self_pos.ne'
+
+variable (R₀ R n e) in
+/-- The statement we induct on in the `C : R → R[X]` case of Chevalley's theorem with complexity
+bound. -/
+private def Statement [Algebra ℤ R] : Prop :=
+  ∀ f : R[X], ∃ T : ConstructibleSetData R,
+    comap Polynomial.C '' (zeroLocus (Set.range e) \ zeroLocus {f}) = T.toSet ∧
+    ∀ C ∈ T, C.n ≤ e.degBound ∧ ∀ i, C.g i ∈ e.coeffSubmodule R₀ ^ e.powBound
+
+end InductionObj
+
+open InductionObj
+
+universe u
+
+/--
+The structure of the induction in the proof of Chevalley's theorem:
+Consider a property on a vector `e` of polynomials. Suppose that it holds for the following cases:
+1. The vector contains a single monic polynomial (and zero otherwise).
+2. The vector contains zeroes only.
+3. Suppose `eᵢ` has the lowest degree among all monic polynomials and `eⱼ` is some other polynomial.
+  If the property holds when `eⱼ` is replaced by `eⱼ % eᵢ`, then it holds for `e`.
+4. Suppose the property holds for both the localization at some leading coefficient of `eᵢ` and
+  the localization at the leading coefficient of `eᵢ`, then the property holds for `e`.
+
+Then it holds for all vectors `e` over all rings.
+-/
+private lemma induction_structure (n : ℕ)
+    (P : ∀ (R : Type u) [CommRing R], (InductionObj R n) → Prop)
+    (hP₀ : ∀ (R) [CommRing R] (e : InductionObj R n) (i : Fin n),
+      (e.1 i).Monic → (∀ j ≠ i, e.1 j = 0) → P R e)
+    (hP₁ : ∀ (R) [CommRing R], P R ⟨0⟩)
+    (hP₃ : ∀ (R) [CommRing R] (e : InductionObj R n) (i j : Fin n),
+      (e.1 i).Monic → (e.1 i).degree ≤ (e.1 j).degree → i ≠ j →
+      P R ⟨update e.1 j (e.1 j %ₘ e.1 i)⟩ → P R e)
+    (hP : ∀ (R) [CommRing R] (c : R) (i : Fin n) (e : InductionObj R n), c = (e.1 i).leadingCoeff →
+      c ≠ 0 →
+      P (Away c) ⟨Polynomial.C (IsLocalization.Away.invSelf (S := Away c) c) •
+        mapRingHom (algebraMap _ _) ∘ e⟩ →
+      P (R ⧸ Ideal.span {c}) ⟨mapRingHom (algebraMap _ _) ∘ e⟩ → P R e)
```

Comment:
Please make the subscripts on the `hP` assumptions match the 1,2,3,4 in the docstring.